src/NNCache.h

```diff
@@ -0,0 +1,72 @@
+/*
+    This file is part of Leela Zero.
+    Copyright (C) 2017 Gian-Carlo Pascutto
+
+    Leela Zero is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Leela Zero is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Leela Zero.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef NNCACHE_H_INCLUDED
+#define NNCACHE_H_INCLUDED
+
+#include "config.h"
+
+#include <vector>
+#include <deque>
+#include <unordered_map>
+
+#include "Network.h"
+#include "SMP.h"
+
+class NNCache {
+public:
+    // return the global NNCache
+    static NNCache* get_NNCache(void);
+
+    // Try and find an existing entry.
+    const Network::Netresult* lookup(const Network::NNPlanes& features);
+
+    // Insert a new entry.
+    void insert(const Network::NNPlanes& features, const Network::Netresult& result);
+
+    // Return the hit rate ratio.
+    std::pair<int, int> hit_rate() const { return {m_hits, m_lookups}; }
+
+    void dump_stats();
+
+private:
+    NNCache(int size = 50000);  // ~ 200MB
+
+    SMP::Mutex m_mutex;
+
+    size_t m_size;
+
+    // Statistics
+    int m_hits{0};
+    int m_lookups{0};
+    int m_inserts{0};
+    int m_collisions{0};
+
+    struct Entry {
+        Entry(const Network::NNPlanes& f, const Network::Netresult& r) : features(f), result(r) {}
+        Network::NNPlanes features; // ~ 1KB
+        Network::Netresult result;  // ~ 3KB
+    };
+
+    // Map from hash to {features, result}
+    std::unordered_map<size_t, std::unique_ptr<const Entry>> m_cache;
+    // Order entries were added to the map.
+    std::deque<size_t> m_order;
```

Comment:
I think you might be able to use a normal queue

Reply:
Amusingly, std::queue is normally implemented on top of std::deque.

Reply:
:)